File: packages/asdl-core/src/asdl_core/gt/metadata_reader.py
"""Graphite metadata-store reader for stack snapshots."""

from __future__ import annotations

import json
import sqlite3
from dataclasses import dataclass
from pathlib import Path

from asdl_core.gt.types import (
    ChildrenCorruption,
    DescendantWalk,
    GtCommandFailure,
    StackFork,
    StackInfo,
    TrunkMarkerClean,
    TrunkMarkerProblem,
    TrunkMarkerStatus,
    UntrackedBranch,
    WalkCompleted,
    WalkCycle,
    WalkRowMissing,
    WalkTermination,
)


@dataclass(frozen=True)
class _BranchMetadataRow:
    parent_branch_name: str | None
    children: tuple[str, ...]
    validation_result: str | None
    children_corruption: ChildrenCorruption | None


_REQUIRED_BRANCH_METADATA_COLUMNS = frozenset(
    {"branch_name", "parent_branch_name", "children", "validation_result"}
)


def _parse_children(
    branch_name: str,
    raw_children: object,
) -> tuple[tuple[str, ...], ChildrenCorruption | None]:
    if raw_children is None:
        return (), None
    if not isinstance(raw_children, str):
        return (), ChildrenCorruption(branch=branch_name, kind="not_text")
    if not raw_children:
        return (), None

    try:
        parsed = json.loads(raw_children)
    except json.JSONDecodeError:
        return (), ChildrenCorruption(branch=branch_name, kind="invalid_json")

    if not isinstance(parsed, list):
        return (), ChildrenCorruption(branch=branch_name, kind="not_list")

    children = tuple(child for child in parsed if isinstance(child, str))
    if len(children) != len(parsed):
        return children, ChildrenCorruption(branch=branch_name, kind="non_string")
    return children, None


def _metadata_text(value: object) -> str | None:
    if not isinstance(value, str):
        return None
    if value == "":
        return None
    return value


def _load_branch_metadata(
    db_path: Path,
) -> tuple[dict[str, _BranchMetadataRow], int] | GtCommandFailure:
    try:
        with sqlite3.connect(f"file:{db_path}?mode=ro", uri=True) as connection:
            table_info = connection.execute("PRAGMA table_info(branch_metadata)").fetchall()
            columns = {record[1] for record in table_info if isinstance(record[1], str)}
            missing_required = _REQUIRED_BRANCH_METADATA_COLUMNS.difference(columns)
            if missing_required:
                return GtCommandFailure(
                    message=(
                        "Graphite metadata schema mismatch: branch_metadata missing "
                        f"required column {sorted(missing_required)[0]}"
                    ),
                    returncode=None,
                )

            records = connection.execute(
                """
                SELECT branch_name, parent_branch_name, children, validation_result
                FROM branch_metadata
                """
            ).fetchall()
    except sqlite3.OperationalError as exc:
        message = str(exc)
        if message.startswith("no such table") or message.startswith("no such column"):
            return GtCommandFailure(
                message=f"Graphite metadata schema mismatch: {exc}",
                returncode=None,
            )
        return GtCommandFailure(
            message=f"Graphite metadata store unreadable: {exc}",
            returncode=None,
        )
    except sqlite3.DatabaseError as exc:
        return GtCommandFailure(
            message=f"Graphite metadata store unreadable: {exc}",
            returncode=None,
        )

    rows: dict[str, _BranchMetadataRow] = {}
    empty_branch_name_rows = 0
    for record in records:
        branch_name, parent_branch_name, raw_children, validation_result = record
        if not isinstance(branch_name, str) or not branch_name:
            empty_branch_name_rows += 1
            continue
        parent = _metadata_text(parent_branch_name)
        validation = _metadata_text(validation_result)
        children, children_corruption = _parse_children(branch_name, raw_children)
        rows[branch_name] = _BranchMetadataRow(
            parent_branch_name=parent,
            children=children,
            validation_result=validation,
            children_corruption=children_corruption,
        )
    return rows, empty_branch_name_rows


def _walk_ancestors(
    rows: dict[str, _BranchMetadataRow],
    current_branch: str,
) -> tuple[tuple[str, ...], str, WalkTermination]:
    ancestor_names_reversed: list[str] = []
    branch = current_branch
    visited = {current_branch}

    while True:
        row = rows[branch]
        parent = row.parent_branch_name
        if parent is None:
            return tuple(reversed(ancestor_names_reversed)), branch, WalkCompleted()
        if parent in visited:
            return tuple(reversed(ancestor_names_reversed)), branch, WalkCycle(parent)

        ancestor_names_reversed.append(parent)
        if parent not in rows:
            return tuple(reversed(ancestor_names_reversed)), parent, WalkRowMissing(parent)

        visited.add(parent)
        branch = parent


def _walk_first_child_descendants(
    rows: dict[str, _BranchMetadataRow],
    current_branch: str,
) -> tuple[tuple[str, ...], DescendantWalk]:
    descendants: list[str] = []
    forks: list[StackFork] = []
    children_corruptions: list[ChildrenCorruption] = []
    branch = current_branch
    visited = {current_branch}

    while True:
        row = rows[branch]
        if row.children_corruption is not None:
            children_corruptions.append(row.children_corruption)
        if len(row.children) > 1:
            forks.append(StackFork(branch=branch, children=row.children))
        if not row.children:
            return tuple(descendants), DescendantWalk(
                forks=tuple(forks),
                children_corruptions=tuple(children_corruptions),
                termination=WalkCompleted(),
            )

        child = row.children[0]
        if child in visited:
            return tuple(descendants), DescendantWalk(
                forks=tuple(forks),
                children_corruptions=tuple(children_corruptions),
                termination=WalkCycle(child),
            )

        descendants.append(child)
        if child not in rows:
            return tuple(descendants), DescendantWalk(
                forks=tuple(forks),
                children_corruptions=tuple(children_corruptions),
                termination=WalkRowMissing(child),
            )

        visited.add(child)
        branch = child


def _trunk_marker_status(
    rows: dict[str, _BranchMetadataRow],
    terminus_branch: str,
) -> TrunkMarkerStatus:
    marked_trunks = tuple(
        branch for branch, row in rows.items() if row.validation_result == "TRUNK"
    )
    if terminus_branch not in rows:
        return TrunkMarkerProblem(
            terminus=terminus_branch,
            terminus_state="row_missing",
            marked_trunks=marked_trunks,
        )

    terminus_state = "marked" if rows[terminus_branch].validation_result == "TRUNK" else "unmarked"
    if terminus_state == "marked" and marked_trunks == (terminus_branch,):
        return TrunkMarkerClean()
    return TrunkMarkerProblem(
        terminus=terminus_branch,
        terminus_state=terminus_state,
        marked_trunks=marked_trunks,
    )


def read_stack_from_metadata_db(
    db_path: Path,
    current_branch: str,
) -> StackInfo | UntrackedBranch | GtCommandFailure:
    """Read the current Graphite stack snapshot from Graphite's SQLite store."""
    if not db_path.exists():
        return GtCommandFailure(
            message=f"Graphite metadata store not found at {db_path}",
            returncode=None,
        )

    loaded = _load_branch_metadata(db_path)
    if isinstance(loaded, GtCommandFailure):
        return loaded
    rows, empty_branch_name_rows = loaded

    if current_branch not in rows:
        return UntrackedBranch(
            message=f"current branch is not tracked by Graphite: {current_branch}"
        )

    ancestors, terminus_branch, ancestor_termination = _walk_ancestors(rows, current_branch)
    descendants, descendant_walk = _walk_first_child_descendants(rows, current_branch)
    trunk_marker = _trunk_marker_status(rows, terminus_branch)

    consumed_corruption_branches = {
        corruption.branch for corruption in descendant_walk.children_corruptions
    }
    unwalked_children_corruptions = tuple(
        row.children_corruption
        for row in rows.values()
        if row.children_corruption is not None
        and row.children_corruption.branch not in consumed_corruption_branches
    )
    return StackInfo(
        trunk=ancestors[0] if ancestors else current_branch,
        current=current_branch,
        ancestors=ancestors,
        children=rows[current_branch].children,
        descendants=descendants,
        ancestor_termination=ancestor_termination,
        descendant_walk=descendant_walk,
        trunk_marker=trunk_marker,
        unwalked_children_corruptions=unwalked_children_corruptions,
        empty_branch_name_rows=empty_branch_name_rows,
    )

File: ts/packages/slot/src/gt/metadata-reader.ts
import { existsSync } from "node:fs";

import { DatabaseSync } from "node:sqlite";

import { commandFailure, type BranchMetadataGraph, type BranchMetadataGraphRow, type ChildrenCorruption, type DescendantWalk, type GtCommandFailure, type StackFork, type StackInfo, type TrunkMarkerStatus, type UntrackedBranch, type WalkTermination } from "./types.ts";

const REQUIRED_COLUMNS = new Set(["branch_name", "parent_branch_name", "children", "validation_result"]);

type SqliteValue = string | number | bigint | Uint8Array | null;
type SqliteRecord = Record<string, SqliteValue>;

export function readBranchGraphFromMetadataDb(dbPath: string): BranchMetadataGraph | GtCommandFailure {
	if (!existsSync(dbPath)) return commandFailure(`Graphite metadata store not found at ${dbPath}`);
	const loaded = loadBranchMetadata(dbPath);
	if ("message" in loaded) return loaded;
	return { rows: [...loaded.rows.values()].sort((left, right) => left.name.localeCompare(right.name)), empty_branch_name_rows: loaded.emptyBranchNameRows };
}

export function readStackFromMetadataDb(dbPath: string, currentBranch: string): StackInfo | UntrackedBranch | GtCommandFailure {
	if (!existsSync(dbPath)) return commandFailure(`Graphite metadata store not found at ${dbPath}`);
	const loaded = loadBranchMetadata(dbPath);
	if ("message" in loaded) return loaded;
	if (!loaded.rows.has(currentBranch)) return { type: "untracked_branch", message: `current branch is not tracked by Graphite: ${currentBranch}` };
	const [ancestors, terminusBranch, ancestorTermination] = walkAncestors(loaded.rows, currentBranch);
	const [descendants, descendantWalk] = walkFirstChildDescendants(loaded.rows, currentBranch);
	const consumed = new Set(descendantWalk.children_corruptions.map((corruption) => corruption.branch));
	const unwalked = [...loaded.rows.values()].flatMap((row) => row.children_corruption !== null && !consumed.has(row.children_corruption.branch) ? [row.children_corruption] : []);
	return {
		trunk: ancestors[0] ?? currentBranch,
		current: currentBranch,
		ancestors,
		children: loaded.rows.get(currentBranch)?.children ?? [],
		descendants,
		ancestor_termination: ancestorTermination,
		descendant_walk: descendantWalk,
		trunk_marker: trunkMarkerStatus(loaded.rows, terminusBranch),
		unwalked_children_corruptions: unwalked,
		empty_branch_name_rows: loaded.emptyBranchNameRows,
	};
}

function loadBranchMetadata(dbPath: string): { rows: Map<string, BranchMetadataGraphRow>; emptyBranchNameRows: number } | GtCommandFailure {
	let db: DatabaseSync;
	try {
		db = new DatabaseSync(dbPath, { readOnly: true });
	} catch (error) {
		return commandFailure(`Graphite metadata store unreadable: ${errorMessage(error)}`);
	}
	try {
		const tableInfo = db.prepare("PRAGMA table_info(branch_metadata)").all() as SqliteRecord[];
		const columns = new Set(tableInfo.map((record) => record.name).filter((name): name is string => typeof name === "string"));
		for (const required of REQUIRED_COLUMNS) {
			if (!columns.has(required)) return commandFailure(`Graphite metadata schema mismatch: branch_metadata missing required column ${required}`);
		}
		const records = db.prepare("SELECT branch_name, parent_branch_name, children, validation_result FROM branch_metadata").all() as SqliteRecord[];
		const rows = new Map<string, BranchMetadataGraphRow>();
		let emptyBranchNameRows = 0;
		for (const record of records) {
			const branchName = record.branch_name;
			if (typeof branchName !== "string" || branchName.length === 0) {
				emptyBranchNameRows += 1;
				continue;
			}
			const parsed = parseChildren(branchName, record.children ?? null);
			rows.set(branchName, {
				name: branchName,
				parent: metadataText(record.parent_branch_name ?? null),
				children: parsed.children,
				validation_result: metadataText(record.validation_result ?? null),
				children_corruption: parsed.corruption,
			});
		}
		return { rows, emptyBranchNameRows };
	} catch (error) {
		const message = errorMessage(error);
		if (message.startsWith("no such table") || message.startsWith("no such column")) return commandFailure(`Graphite metadata schema mismatch: ${message}`);
		return commandFailure(`Graphite metadata store unreadable: ${message}`);
	} finally {
		db.close();
	}
}

function parseChildren(branchName: string, rawChildren: SqliteValue): { children: readonly string[]; corruption: ChildrenCorruption | null } {
	if (rawChildren === null) return { children: [], corruption: null };
	if (typeof rawChildren !== "string") return { children: [], corruption: { branch: branchName, kind: "not_text" } };
	if (rawChildren.length === 0) return { children: [], corruption: null };
	let parsed: unknown;
	try {
		parsed = JSON.parse(rawChildren);
	} catch {
		return { children: [], corruption: { branch: branchName, kind: "invalid_json" } };
	}
	if (!Array.isArray(parsed)) return { children: [], corruption: { branch: branchName, kind: "not_list" } };
	const children = parsed.filter((child): child is string => typeof child === "string");
	return { children, corruption: children.length === parsed.length ? null : { branch: branchName, kind: "non_string" } };
}

function metadataText(value: SqliteValue): string | null {
	return typeof value === "string" && value.length > 0 ? value : null;
}

function walkAncestors(rows: Map<string, BranchMetadataGraphRow>, currentBranch: string): [readonly string[], string, WalkTermination] {
	const reversed: string[] = [];
	let branch = currentBranch;
	const visited = new Set([currentBranch]);
	while (true) {
		const row = rows.get(branch);
		if (row === undefined || row.parent === null) return [reversed.reverse(), branch, { type: "completed" }];
		if (visited.has(row.parent)) return [reversed.reverse(), branch, { type: "cycle", branch: row.parent }];
		reversed.push(row.parent);
		if (!rows.has(row.parent)) return [reversed.reverse(), row.parent, { type: "row_missing", branch: row.parent }];
		visited.add(row.parent);
		branch = row.parent;
	}
}

function walkFirstChildDescendants(rows: Map<string, BranchMetadataGraphRow>, currentBranch: string): [readonly string[], DescendantWalk] {
	const descendants: string[] = [];
	const forks: StackFork[] = [];
	const corruptions: ChildrenCorruption[] = [];
	let branch = currentBranch;
	const visited = new Set([currentBranch]);
	while (true) {
		const row = rows.get(branch);
		if (row === undefined) return [descendants, { forks, children_corruptions: corruptions, termination: { type: "row_missing", branch } }];
		if (row.children_corruption !== null) corruptions.push(row.children_corruption);
		if (row.children.length > 1) forks.push({ branch, children: row.children });
		const child = row.children[0];
		if (child === undefined) return [descendants, { forks, children_corruptions: corruptions, termination: { type: "completed" } }];
		if (visited.has(child)) return [descendants, { forks, children_corruptions: corruptions, termination: { type: "cycle", branch: child } }];
		descendants.push(child);
		if (!rows.has(child)) return [descendants, { forks, children_corruptions: corruptions, termination: { type: "row_missing", branch: child } }];
		visited.add(child);
		branch = child;
	}
}

function trunkMarkerStatus(rows: Map<string, BranchMetadataGraphRow>, terminusBranch: string): TrunkMarkerStatus {
	const markedTrunks = [...rows.values()].filter((row) => row.validation_result === "TRUNK").map((row) => row.name);
	const terminus = rows.get(terminusBranch);
	if (terminus === undefined) return { type: "problem", terminus: terminusBranch, terminus_state: "row_missing", marked_trunks: markedTrunks };
	const terminusState = terminus.validation_result === "TRUNK" ? "marked" : "unmarked";
	if (terminusState === "marked" && markedTrunks.length === 1 && markedTrunks[0] === terminusBranch) return { type: "clean" };
	return { type: "problem", terminus: terminusBranch, terminus_state: terminusState, marked_trunks: markedTrunks };
}

function errorMessage(error: unknown): string {
	return error instanceof Error ? error.message : String(error);
}
